imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+            for i in range(0, len(uncompressed_vectors), uncompressed_vector_size):
+                decoded_vector = MagL1a.unpack_one_vector(
+                    uncompressed_vectors[i : i + uncompressed_vector_size],
+                    compression_width,
+                    False,
+                )
+                primary_vectors[vector_index] = decoded_vector
+                primary_vectors[vector_index][3] = first_vector[3]
+                vector_index += 1
+
+        # Secondary vector processing
+        first_secondary_vector = MagL1a.unpack_one_vector(
+            vector_bits[
+                end_primary_vector : end_primary_vector + uncompressed_vector_size + 2
+            ],
+            compression_width,
+            True,
+        )
+
+        # Split up the bit array, skipping past the primary vector and uncompressed
+        # starting vector
+        secondary_split_bits = np.split(
+            vector_bits[: secondary_boundaries[-1]], secondary_boundaries[:-1]
+        )[1:]
+
+        vector_diffs = list(map(MagL1a.decode_fib_zig_zag, secondary_split_bits))
+
+        secondary_vectors = MagL1a.accumulate_vectors(
+            first_secondary_vector, vector_diffs, secondary_count
+        )
+        # print(f"Secondary vectors: {secondary_vectors}")
+
+        secondary_vector_missing = secondary_count - math.ceil(len(secondary_split_bits) / 3) - 1
+
+        secondary_vector_missing = secondary_count - len(secondary_split_bits) // 3 - 1
+        if secondary_vector_missing:
+            vector_index = secondary_count - secondary_vector_missing
+            secondary_end = (
+                secondary_boundaries[-1]
+                + uncompressed_vector_size * secondary_vector_missing
+            )
+            uncompressed_vectors = vector_bits[secondary_boundaries[-1] : secondary_end]
+            for i in range(0, len(uncompressed_vectors), uncompressed_vector_size):
+                decoded_vector = MagL1a.unpack_one_vector(
+                    uncompressed_vectors[i : i + uncompressed_vector_size],
+                    compression_width,
+                    False,
+                )
+                secondary_vectors[vector_index] = decoded_vector
+                secondary_vectors[vector_index][3] = first_secondary_vector[3]
+                vector_index += 1
+
+        # If there is a range data section, it describes all the data, compressed or
+        # uncompressed.
+        if has_range_data_section:
+            primary_vectors = MagL1a.process_range_data_section(
+                bit_array[end_vector : end_vector + (primary_count - 1) * 2],
+                primary_vectors,
+            )
+            secondary_vectors = MagL1a.process_range_data_section(
+                bit_array[
+                    end_vector + (primary_count - 1) * 2 : end_vector
+                    + (primary_count + secondary_count - 2) * 2
+                ],
+                secondary_vectors,
+            )
```

Comment:
Not really clear from this that if `has_range_data_section=0` then all vectors have the same range as their sensors first vector. Refactor the processing of the range section out to a method and make that logic explicit?

Reply:
So, do you mean that I should always call `process_range_data_section` and pass `has_range_data_section` through, or should I just be more explicit within this if statement?

Reply:
I think maybe calling a method called `unpack_vector_ranges` that either copies it from the first vector, or reads the end section would be more explicit/easier to understand. ATM there are 2 really separate bits of code, (line 770 `vectors[i + compressed_count][3] = vectors[0][3]`) and here that generate the range for each vector, depending on the value of `has_range_data_section` and it is non-obvious that they are connected

Reply:
Ok yeah, that makes sense. I will rework that, I think it will be more clear. Thanks!